packages/x-date-pickers/src/internals/hooks/usePickerState.ts

```diff
@@ -41,127 +79,162 @@ interface PickerStateProps<TInput, TDateValue> {
-    dispatch({ type: 'reset', payload: parsedDateValue });
-  }
-
-  const [initialDate, setInitialDate] = React.useState<TDateValue>(draftState.committed);
+  const [dateState, setDateState] = React.useState<DateState<TDate>>(() => ({
+    committed: parsedDateValue,
+    draft: parsedDateValue,
+    resetTarget: parsedDateValue,
+  }));
 
-  // Mobile keyboard view is a special case.
-  // When it's open picker should work like closed, because we are just showing text field
-  const [isMobileKeyboardViewOpen, setMobileKeyboardViewOpen] = React.useState(false);
+  const setDate = React.useCallback(
+    (params: DateStateAction<TDate>) => {
+      setDateState((prev) => {
+        switch (params.action) {
+          case 'setAll':
+          case 'acceptAndClose': {
+            return { draft: params.value, committed: params.value, resetTarget: params.value };
+          }
+          case 'setCommitted': {
+            return { ...prev, draft: params.value, committed: params.value };
+          }
+          case 'setDraft': {
+            return { ...prev, draft: params.value };
+          }
+          default: {
+            return prev;
+          }
+        }
+      });
 
-  const acceptDate = React.useCallback(
-    (acceptedDate: TDateValue, needClosePicker: boolean) => {
-      onChange(acceptedDate);
+      if (!params.internal) {
+        onChange(params.value);
+      }
 
-      if (needClosePicker) {
+      if (params.action === 'acceptAndClose') {
         setIsOpen(false);
-        setInitialDate(acceptedDate);
         if (onAccept) {
-          onAccept(acceptedDate);
+          onAccept(params.value);
         }
       }
     },
     [onAccept, onChange, setIsOpen],
   );
 
+  React.useEffect(() => {
+    if (parsedDateValue != null) {
+      setLastValidDateValue(parsedDateValue);
+    }
+  }, [parsedDateValue]);
+
+  // Set the draft and committed date to equal the new prop value.
+  if (!valueManager.areValuesEqual(utils, dateState.committed, parsedDateValue)) {
+    setDate({ action: 'setCommitted', value: parsedDateValue, internal: true });
+  }
+
+  const shouldCloseOnSelect = !(disableCloseOnSelect ?? wrapperVariant === 'mobile');
+
   const wrapperProps = React.useMemo(
     () => ({
       open: isOpen,
-      onClear: () => acceptDate(valueManager.emptyValue, true),
-      onAccept: () => acceptDate(draftState.draft, true),
-      onDismiss: () => acceptDate(initialDate, true),
+      onClear: () => {
+        // Reset all date in state to the empty value and close picker
+        setDate({ value: valueManager.emptyValue, action: 'acceptAndClose' });
+      },
+      onAccept: () => {
+        // Set all date in state to equal the current draft value and close picker
+        setDate({ value: dateState.draft, action: 'acceptAndClose' });
+      },
+      onDismiss: () => {
+        if (shouldCloseOnSelect) {
+          // Set all dates in state to equal the last accepted date
```

Comment:
```suggestion
          // Closing input before finishing to enter the date reset input value
```

Reply:
"Closing input" ?

The naming I see is

- Input => the actual TextField
- View => the "menu" opened when we click on the icon

Reply:
So "Closing view".
I'm still confused about the distinction between the TextFields and Views management

My point is that this `if` condition is not straight forward, because as a naive reader I would expect a `shouldSelectOnClose` instead of the `shouldCloseOnSelect`. So the comment should give context about when the `onDismiss` is used, and why it depends on `shouldCloseOnSelect`

Reply:
> I'm still confused about the distinction between the TextFields and Views management

Same
This RFC https://github.com/mui/mui-x/issues/4484 could help us clarify the naming.

____

I think we can distinguish two things:

- The documentation of the methods `onDismiss`, `onSetToday` etc... which have zero JSDoc, are defined in `PickersModalDialogProps` / `PrivateWrapperProps` and `wrapperProps`

- The comment inside those methods

My comments were more about internal mixture.
But we should probably add a clean documentation that could even be used on the doc at some point.